README.TXT
ISO Master
=======

A graphical CD/DVD image editor for Linux and BSD systems.

[English]

Overview
----------
ISO Master is an open-source, easy to use graphical CD image editor. You can use this 
program to:
- Add files to an ISO image
- Extract files from an ISO image
- Delete files from an ISO image
- Create bootable ISOs

Supported formats:
- Read: ISO, NRG, MDF
- Write: ISO only

Requirements
------------
- GTK2 development libraries
- Standard build tools (gcc, make)

Compiling
--------
  make

Installing (optional)
-------------------
  sudo make install

Or on Ubuntu/Debian:
  sudo make install

Usage
-----
1. Open an ISO file: File -> Open (Ctrl+O)
2. Add files: Select files from the left pane, click "Add" button
3. Extract files: Select files in the ISO, click "Extract" button
4. Delete files: Select files, press Delete key
5. Save: File -> Save (directly to current file) 
   or File -> Save As (save as new file)

Keyboard shortcuts:
- Ctrl+N: New ISO
- Ctrl+O: Open ISO
- Ctrl+S: Save
- Ctrl+Shift+S: Save As
- F2: Rename
- F3: View
- F4: Edit
- F5: Refresh
- Ctrl+Q: Quit

License
-------
GNU General Public License version 2. See LICENCE.TXT for full text.

--------------------------------------------------------------------

ISO Master
========

一款适用于 Linux 和 BSD 系统的图形化 CD/DVD 镜像编辑器。

简介
------
ISO Master 是一款开源、易于使用的图形化 CD 镜像编辑器。您可以使用此程序：
- 向 ISO 镜像添加文件
- 从 ISO 镜像提取文件
- 从 ISO 镜像删除文件
- 创建可启动 ISO

支持的格式：
- 读取：ISO、NRG、MDF
- 写入：仅支持 ISO

编译要求
--------
- GTK2 开发库
- 标准编译工具（gcc、make）

编译方法
--------
  make

安装（可选）
-----------
  sudo make install

或在 Ubuntu/Debian 系统上：
  sudo make install

使用方法
--------
1. 打开 ISO 文件：文件 -> 打开（Ctrl+O）
2. 添加文件：从左侧窗格选择文件，点击"添加"按钮
3. 提取文件：在 ISO 中选择文件，点击"提取"按钮
4. 删除文件：选择文件，按 Delete 键
5. 保存：文件 -> 保存（直接保存到当前文件）
   或 文件 -> 另存为（保存为新文件）

快捷键：
- Ctrl+N：新建 ISO
- Ctrl+O：打开 ISO
- Ctrl+S：保存
- Ctrl+Shift+S：另存为
- F2：重命名
- F3：查看
- F4：编辑
- F5：刷新
- Ctrl+Q：退出

许可证
------
GNU 通用公共许可证版本 2。详见 LICENCE.TXT。

CHANGELOG.TXT
4 Jan 2009 - 1.3.7
==================
- Fixed a bug where the duplicate scan would cause a corrupted image if files of teh same size were added from the filesystem
- Fixed two minor errors that most people probably have never seen

1 Sep 2009 - 1.3.6
==================
- Added an 'open recent' function.
- Added Latvian translation.
- Fixed some error messages (unknown error was displayed before).
- Target of dead symbolic links is no longer overwritten without a prompt.
- Documentation for bkisofs is now installed together with the program.

30 Dec 2008 - 1.3.5
===================
- Updated Danish and Chinese (Taiwan) translations.
- Updated .desktop file to be compatible with current spec.
- Fixed some error messages (unknown error was displayed before).
- Target of dead symbolic links is no longer overwritten without a prompt.
- Documentation for bkisofs is now installed together with the program.
- Added the 'Recently opened' function.

29 Nov 2008 - 1.3.4
===================
- Updated Italian, Portuguese, Serbian Latin, and Spanish translations.
- Added arabic translation.

30 Jun 2008 - 1.3.3
===================
- Added Portuguese (Brasil) translation.

28 Jun 2008 - 1.3.2
===================
- Updated Catalan, Italian, Dutch, and Faroese translations.
- Added Danish translation.
- Fixed about window to show the program name and version properly.
- Added a dummy configure script.
- Fixed bug where filenames would have ';1' terminators when only ISO9660 filenames were on the original ISO.

03 Feb 2008 - 1.3.1
===================
- Added Belarusian and Estonian translations.
- Fixed a bug where the boot media type for floppies was read wrong.
- Fixed bug where a file extracted from a very large image would sometimes be corrupt.

17 Dec 2007 - 1.3
=================
- Added Hebrew and Indonesian translations.
- A few usability improvements.
- Sorting no longer has to be case-sensitive.
- Fixed bug where a false positive isolinux detection would cause some ISOs to not be bootable anyy more.
- Can now view and edit files on the regular filesystem, not just on the image.

27 Oct 2007 - 1.2
=================
- Added Faroese, Slovenian, Hungarian translations.
- Fixed bug where El Torito boot records with size not divisible by 4 were not allowed.
- Progress bars are now much smoother.
- Name of the open file is now in the title ofthe window.
- If the viewer or editor fails to start, an error message is now displayed.

27 Aug 2007 - 1.1
=================
- Added Serbian/Latin, and Turkish translations.
- Fixed memory leak when a directory is reloaded, very serious if the directory has many children.
- Can now view files with an app the user can set.
- Can now edit files with an app the user can set.
- Some MDF files can now be read.
- Several small usability improvements.
- Showing the contents of a directory now takes much less time.
- Fixed bug where the symlinks written would be interpreted as invalid by some systems.
- Improved sorting by size (directories always first now works).
- Fixed some typos in the GUI.
- Added right-click > rename menu.
- Mapped F3 to view, F4 to edit, and F5 to refresh.
- Can now change the permissions of files and directories.
- Sort column is now remembered between sessions.
- Canceling an operation now works even if several rows are selected.

01 Jun 2007 - 1.0
=================
- Fixed bug where extract button generated gtk error when no rows selected.
- Creating or adding "." and ".." no longer allowed.
- Added Ukrainian, Serbian, Croation, Catalan, Swedish, Norwegian (Bokmal and Nynorsk), Finnish, Czech, and Greek translations.
- Added checkbox to append an extension to a saved filename automatically.
- Now less memory/CPU is used when saving an image.
- Can now compile on distros with older GTK versions.
- Can now rename files and directories on the iso.
- Small usability improvements.
- Fixed bug where a saved image with symbolic links would end up having some corrupted files.

17 Mar 2007 - 0.8
=================
- Added 3 more icons by kearone for a consistent look.
- Tested on big-endian machine, confirmed everything works as it should.
- Fixed allignment problem on some strange CPUs.
- Added Spanish, Italian, Slovak, Simplified Chineese, Romanian, Dutch, Portuguese translations.
- Added isomaster.desktop file (used as a shortcut in desktop environments).
- Added support for symbolic links (both on the ISO and the filesystem).
- Added support for duplicate files (a.k.a. hard links).
- Several usability improvements.

11 Jan 2007 - 0.7
=================
- Fixed the double warning message when adding directories.
- Changed bkisofs to make it ready for links and get rid of a lot of duplicate code.
- Added Bulgarian, Chinese, Russian, German, French, Polish and Albanian translations.
- Fixed 'bug' that images with a broken root directory identifier would fail to open.
- Writing and extracting operations can now be canceled.
- Minor usability improvements.
- Can now open and read (but not write) .nrg images.
- Writing progress bar now works in percentage mode instead of activity mode.
- Images that failed to write fully are now deleted.
- Added Help/Overview with a brief description of how to use the program.
- Changed mangling so that winxp CDs would still boot.
- Tested on a 64bit machine, confirmed everything works as it should.
- Now remember the scroll offset for the browsers so adding/extracting/deleting something doesn't cause the browser to scroll to the top.

09 Dec 2006 - 0.6
=================
- Replaced the Add and Extract icons with the two I got from kearone.
- Added gettext() support so that can now make use of .po files.
- Can now read, add, extract and write files with long directory names (up to 255 characters for RockRidge).
- Am allowing more characters in filenames on the image.
- Enter key now works in 'reate new directory' dialog.
- Added button to refresh the filesystem and iso view.

about.c
/******************************* LICENCE **************************************
* Any code in this file may be redistributed or modified under the terms of
* the GNU General Public Licence as published by the Free Software 
* Foundation; version 2 of the licence.
****************************** END LICENCE ***********************************/

/******************************************************************************
* Author:
* Andrew Smith, http://littlesvr.ca/misc/contactandrew.php
*
* Contributors:
* 
******************************************************************************/

#include <gtk/gtk.h>
#include <gdk/gdkkeysyms.h>

#include "isomaster.h"

extern GtkWidget* GBLmainWindow;

#if GTK_MINOR_VERSION >= 6

static const char* 
GBLprogramName = "ISO Master "VERSION;

static const char* 
GBLauthors[2] = {
"Many thanks to all the following people:\n"
"\n"
"Andrew Smith\n"
"http://littlesvr.ca/misc/contactandrew.php\n"
"Summer 2005 - Summer 2007\n"
"- author and maintainer\n"
"\n"
"Barb Czegel\n"
"http://cs.senecac.on.ca/~barb.czegel/\n"
"Summer-fall 2005\n"
"- one of my teachers from Seneca College who has been kind enough to let me\n"
"  work on ISO Master as my systems project for two semesters\n"
"\n"
"Nicolas Devillard\n"
"http://ndevilla.free.fr/iniparser/\n"
"August 2006\n"
"- the excellent iniparser, for storing and reading config files\n"
"\n"
"The Samba Project\n"
"http://samba.org/\n"
"December 2005\n"
"- most of the filename mangling code I copied from Samba\n"
"\n"
"Steffen Winterfeldt\n"
"September 2006\n"
"- helped me figure out how to work with isolinux boot records\n"
"\n"
"Roman Hubatsch (kearone)\n"
"http://kearone.deviantart.com/\n"
"December 2006 - January 2007\n"
"- the excellent 'add', 'extract' icons and the new ISO Master icon\n"
"- equally excellent 'go back', 'new folder' and 'delete' icons\n"
"\n"
"David Johnson\n"
"http://www.david-web.co.uk/\n"
"September 2006 - February 2007\n"
"- a patch to allow associating ISO Master with ISO files in file managers\n"
"- a patch to give ISO Master the --help command line parameter\n"
"- gave me access to a 32bit ARM box running Debian for testing\n"
"- gave me access to a big endian SuperSPARC II box running Debian for testing\n"
"\n"
"Ernst W. Winter\n"
"December 2006\n"
"- gave me access to an AMD64 box running OpenBSD for testing\n"
"\n"
"Tango Desktop Project\n"
"http://tango.freedesktop.org/Tango_Icon_Gallery\n"
"August 2006\n"
"- the pretty 'new folder' icon (replaced with kearone's version in 0.8)\n"
"\n"
"Packages:\n"
"\n"
"David Johnson\n"
"http://www.david-web.co.uk/\n"
"- Debian packages of ISO Master, versions 0.1 - 1.2\n"
"- Nexenta GNU/OpenSolaris package of ISO Master, version 1.0\n"
"\n"
"Marcin Zajaczkowski\n"
"http://timeoff.wsisiz.edu.pl/rpms.html\n"
"- Fedora packages of ISO Master, versions 0.3 - 1.2\n"
"\n"
"Toni Graffy\n"
"Maintainer of many SuSE packages at PackMan\n"
"- SuSE packages of ISO Master, versions 0.4 - 1.2\n"
"\n"
"Maciej Libuda\n"
"- Arch packages of ISO Master, versions 0.3 - 1.2\n"
"\n"
"GuestToo\n"
"- Puppy packages of ISO Master, versions 0.1, 0.4 - 1.0\n"
"\n"
"vktgz\n"
"http://www.vktgz.homelinux.net/\n"
"- Gentoo ebuilds of ISO Master, versions 0.4 - 1.2\n"
"\n"
"Babak Farrokhi\n"
"- FreeBSD ports package of ISO Master, versions 0.6 - 1.1\n"
"\n"
"James Bowling\n"
"http://www.jamesbowling.com/\n"
"- Slackware packages of ISO Master, versions 0.4 - 0.5\n"
"\n"
"Michael Shigorin\n"
"- Alt packages of ISO Master, versions 0.5, 0.7, 1.2\n"
"\n"
"Rene Eric\n"
"SLAX packages of versions 0.7 - 1.2\n"
"\n"
"Adam Williamson\n"
"http://www.happyassassin.net/\n"
"- Mandriva package of ISO Master, versions 0.8 - 1.2\n"
"\n"
"Matias A. Fonzo\n"
"- Slackware package of ISO Master, versions 0.8 - 1.3.1\n"
"\n"
"Giovanni Bechis\n"
"http://www.snb.it/\n"
"- OpenBSD port of ISO Master, versions 1.0 - 1.2\n"
"\n"
"Pierrick Le Brun\n"
"- Zenwalk package of ISO Master, versions 1.0 - 1.2\n"
"\n"
"Todd (muggins)\n"
"- Puppy package of ISO Master, versions 1.1 - 1.2\n"
"\n"
"Kenan Pelit\n"
"- Pardus package of ISO Master, version 1.2\n"
"\n"
"F. Lombardi\n"
"SLAX package of ISO Master, version 0.7\n"
"\n"
"Bogdan Radulescu\n"
"- NimbleX package of ISO Master, version 0.7\n"
"\n"
"Wolven\n"
"http://wolvix.org/\n"
"- Wolvix package of ISO Master, version 1.0\n"
"\n"
,
NULL};

static const char* 
GBLtranslators = 
"Abdellah Chelli\n"
"- ar (arabic) translation of ISO Master version 1.3\n"
"\n"
"Ihar Hrachyshka\n"
"- be (Belarusian) translation of ISO Master version 1.3\n"
"\n"
"Dessislav Petrov\n"
"- bg (Bulgarian) translation of ISO Master versions 0.6 - 1.3\n"
"\n"
"Rafael Carreras\n"
"- ca (Catalan) translation of ISO Master version 1.1\n"
"\n"
"Josep Sanchez Mesegue\n"
"- ca (Catalan) translation of ISO Master version 1.2\n"
"\n"
"Schmaki\n"
"- cs (Czech) translation of ISO Master version 1.1\n"
"\n"
"Jiri Huf\n"
"- cs (Czech) translation of ISO Master version 1.3\n"
"\n"
"Toni Graffy\n"
"- de (German) translation of ISO Master versions 0.6 - 1.3\n"
"\n"
"Magnus Valle\n"
"- da (Danish) translation of ISO Master version 1.3\n"
"\n"
"Henrik Kristiansen\n"
"- da (Danish) translation of ISO Master version 1.3\n"
"\n"
"Mike Kranidis\n"
"- el (Greek) translation of ISO Master versions 1.1 - 1.3\n"
"\n"
"Juan Garcia-Murga Monago\n"
"- es (Spanish) translation of ISO Master versions 0.7 - 1.1\n"
"\n"
"Leandro Doctors\n"
"- es (Spanish) translation of ISO Master version 1.3\n"
"\n"
"Christobal\n"
"- es (Spanish) translation of ISO Master version 1.3\n"
"\n"
"Edmund Laugasson\n"
"- et (Estonian) translation of ISO Master version 1.3\n"
"\n"
"Eero Salokannel\n"
"- fi (Finnish) translation of ISO Master versions 1.1 - 1.3\n"
"\n"
"Magnus Valle\n"
"- fo (Faroese) translation of ISO Master version 1.1\n"
"\n"
"Leif Thande\n"
"- fr (French) translation of ISO Master version 0.6\n"
"\n"
"Andrew Smith\n"
"- fr (French) translation of ISO Master version 0.7\n"
"\n"
"Bruno Bord\n"
"- fr (French) translation of ISO Master versions 0.8 - 1.3\n"
"\n"
"Fima Kachinski\n"
"- he (Hebrew) translation of ISO Master version 1.3\n"
"\n"
"Ante Karamatic\n"
"- hr (Croatian) translation of ISO Master version 1.0\n"
"\n"
"Andras Horvath\n"
"- hu (Hungarian) translation of ISO Master version 1.3\n"
"\n"
"Wahyu Puji L\n"
"- id (Indonesian) translation of ISO Master version 1.3\n"
"\n"
"F. Lombardi\n"
"- it (Italian) translation of ISO Master version 0.7 - 1.0\n"
"\n"
"Valerio Guaglianone\n"
"- it (Italian) translation of ISO Master version 0.7, 1.1, 1.3\n"
"\n"
"Rihards Prieditis\n"
"- lv (Latvian) translation of ISO Master version 1.3\n"
"\n"
"Robert Groenning\n"
"- nb (Norwegian Bokmal) translation of ISO Master version 1.0\n"
"\n"
"Stephen Brandt\n"
"- nl (Dutch) translation of ISO Master versions 0.7 - 1.3\n"
"\n"
"Kristof Bal\n"
"- nl (Dutch) translation of ISO Master version 1.3\n"
"\n"
"Robert Groenning\n"
"- nn (Norwegian Nynorsk) translation of ISO Master version 1.1\n"
"\n"
"Maciej Libuda\n"
"- pl (Polish) translation of ISO Master versions 0.6 - 1.3\n"
"\n"
"Joao de Ferro\n"
"- pt (Portuguese) translation of ISO Master versions 0.8, 1.0\n"
"\n"
"Antonio Sousa\n"
"- pt (Portuguese) translation of ISO Master version 1.3\n"
"\n"
"Alexandre Marques\n"
"- pt_BR (Portuguese/Brasil) translation of ISO Master version 1.3\n"
"\n"
"Bogdan Radulescu\n"
"- ro (Romanian) translation of ISO Master version 0.7 - 0.8, 1.0\n"
"\n"
"Anton Obidin\n"
"- ru (Russian) translation of ISO Master versions 0.6 - 1.3\n"
"\n"
"Jozef Riha\n"
"- sk (Slovak) translation of ISO Master versions 1.0 - 1.3\n"
"\n"
"Uros Golob\n"
"- sl (Slovenian) translation of ISO Master versions 1.1 - 1.3\n"
"\n"
"Besnik Bleta\n"
"- sq (Albanian) translation of ISO Master versions 0.8 - 1.3\n"
"\n"
"Danijel Muharemovic\n"
"- sr (Serbian) translation of ISO Master version 1.1\n"
"\n"
"Danijel Muharemovic\n"
"- sr@latin (Serbian/Latin) translation of ISO Master version 1.1\n"
"\n"
"Daniel Nylander\n"
"- sv (Swedish) translation of ISO Master version 1.1\n"
"\n"
"Tutku Dalmaz\n"
"- tr (Turkish) translation of ISO Master version 1.0. 1.3\n"
"\n"
"Serhey Kusyumoff\n"
"- uk (Ukrainian) translation of ISO Master versions 0.8, 1.3\n"
"\n"
"Sun Minchao\n"
"- zh_CN (Simplified Chineese) translation of ISO Master version 0.7 - 1.3\n"
"\n"
"Cheng-Wei Chien\n"
"- zh_TW (Chineese/Taiwan) translation of ISO Master version 0.6 - 1.3\n";

static const char* 
GBLcomments = "An application for editing ISO9660 images based on the "
              "bkisofs access library and the GTK2 GUI toolkit.";

static const char* 
GBLcopyright = "Copyright 2005-2008 Andrew Smith";

static const char* 
GBLwebsite = "http://littlesvr.ca/isomaster/";

#endif /* GTK_MINOR_VERSION >= 6 */

static const char* 
GBLlicense = 
"ISO Master and bkisofs are distributed under the GNU General Public Licence version 2,\n"
"on Debian systems please see /usr/share/common-licenses/GPL-2 for the complete text\n";

static const char*
GBLhelp = 
N_(
"The ISO Master window is split in 2 parts:\n"
" - The top file browser shows files and directories on your machine.\n"
" - The bottom file browser shows files and directories on the ISO image.\n"
"\n"
"To open an existing ISO file click on 'File' and 'Open'. To create a new\n"
"ISO file click on 'File' and 'New'.\n"
"\n"
"To add one or more files or directories to the ISO select them in the top\n"
"file browser and click the 'Add to the ISO' button in the middle toolbar.\n"
"\n"
"To extract one or more files or directories from the ISO select them in\n"
"the bottom file browser and click the 'Extract from ISO' button in the\n"
"middle toolbar.\n"
"\n"
"You can also delete files or directories from the ISO and create new\n"
"directories both on the ISO and your local filesystem.\n"
"\n"
"If you select something in the bottom file browser and right-click on it,\n"
"there are several more options available.\n"
"\n"
"Once you made all the changes to the new ISO image, click on 'File' and\n"
"'Save As'. You cannot overwrite the original ISO.\n"
);

gboolean helpKeyPressCbk(GtkWidget* widget, GdkEventKey* event, gpointer user_data)
{
    if(event->keyval == GDK_Escape)
    {
        gtk_widget_destroy(widget);
        
        return TRUE;
    }
    
    return FALSE;
}

void showAboutWindowCbk(GtkMenuItem* menuItem, gpointer data)
{
#if GTK_MINOR_VERSION >= 6
    gtk_show_about_dialog(GTK_WINDOW(GBLmainWindow), 
                          "name", GBLprogramName,
#if GTK_MINOR_VERSION >= 12
                          "program-name", GBLprogramName,
#endif
                          "authors", GBLauthors,
                          "translator-credits", GBLtranslators,
                          "comments", GBLcomments,
                          "copyright", GBLcopyright,
                          "license", GBLlicense,
                          "website", GBLwebsite,
                          NULL);
#endif
}

/* This callback is also used by an accelerator so make sure you don't use 
* the parameters, since they may not be the menuitem parameters */
void showHelpOverviewCbk(GtkMenuItem* menuItem, gpointer data)
{
    GtkWidget* window;
    GtkWidget* label;
    
    window = gtk_window_new(GTK_WINDOW_TOPLEVEL);
    gtk_window_set_title(GTK_WINDOW(window), _("ISO Master Help"));
    gtk_window_set_transient_for(GTK_WINDOW(window), GTK_WINDOW(GBLmainWindow));
    g_signal_connect(window, "key-press-event", (GCallback)helpKeyPressCbk, NULL);
    
    label = gtk_label_new(_(GBLhelp));
    gtk_container_add(GTK_CONTAINER(window), label);
    gtk_widget_show(label);

    gtk_widget_show(window);
}
